File: ruby/run_tests.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Run all Ruby proxy examples as tests.
#
# Configuration via environment variables:
#     PROXY_URL       - Proxy URL (required), e.g., http://user:pass@proxy:8080
#     TEST_URL        - URL to request (default: https://api.ipify.org?format=json)
#     RESPONSE_HEADER - Optional header name to print from the response
#
# Usage:
#     bundle exec ruby run_tests.rb              # Run all examples
#     bundle exec ruby run_tests.rb httpclient   # Run specific example
#     bundle exec ruby run_tests.rb -l           # List available examples
#
# Install dependencies first: bundle install
require 'open3'
require 'uri'

SCRIPT_DIR = File.expand_path(__dir__)

EXAMPLES = %w[
  net-http-proxy.rb
  httpclient-proxy.rb
  httpclient-scrape-proxy.rb
  httparty-proxy.rb
  typhoeus-proxy.rb
  excon-proxy.rb
].freeze

def parse_args(argv)
  options = { list: false, help: false, examples: [] }
  argv.each do |arg|
    case arg
    when '-l', '--list' then options[:list] = true
    when '-h', '--help' then options[:help] = true
    when /^-/
      # ignore unknown flags
    else
      options[:examples] << arg
    end
  end
  options
end

def show_help
  puts <<~HELP
    Run all Ruby proxy examples as tests.

    Usage:
        bundle exec ruby run_tests.rb [options] [example ...]

    Options:
        -l, --list       List available examples
        -h, --help       Show this help message

    Environment variables:
        PROXY_URL        Proxy URL (required), e.g., http://user:pass@proxy:8080
        TEST_URL         URL to request (default: https://api.ipify.org?format=json)
        RESPONSE_HEADER  Optional header name to print from the response

    Install dependencies first: bundle install
  HELP
end

def list_examples
  puts 'Available examples:'
  EXAMPLES.each do |f|
    puts "  #{f.sub(/-proxy\.rb$/, '')}"
  end
end

def mask_password(url)
  u = URI.parse(url)
  return url unless u.password

  url.sub(":#{u.password}@", ':****@')
rescue URI::InvalidURIError
  url
end

def run_example(filename)
  path = File.join(SCRIPT_DIR, filename)
  return [false, 'script not found'] unless File.file?(path)

  env = ENV.to_h
  _stdout, stderr, status = Open3.capture3(
    env,
    'bundle', 'exec', 'ruby', path,
    chdir: SCRIPT_DIR
  )
  [status.success?, stderr.lines.first&.strip]
end

options = parse_args(ARGV)

if options[:help]
  show_help
  exit 0
end

if options[:list]
  list_examples
  exit 0
end

unless ENV['PROXY_URL'] || ENV['HTTPS_PROXY']
  warn 'Error: Set PROXY_URL environment variable'
  warn ''
  warn 'Example:'
  warn "  export PROXY_URL='http://user:pass@proxy:8080'"
  exit 1
end

to_run = if options[:examples].empty?
           EXAMPLES
         else
           EXAMPLES.select do |f|
             stem = f.sub('-proxy.rb', '')
             options[:examples].any? { |arg| f.include?(arg) || stem.start_with?(arg) }
           end
         end

if to_run.empty?
  warn 'No matching examples.'
  exit 1
end

proxy_url = ENV['PROXY_URL'] || ENV['HTTPS_PROXY']
test_url = ENV['TEST_URL'] || 'https://api.ipify.org?format=json'

puts '=' * 60
puts 'Ruby Proxy Examples - Test Runner'
puts '=' * 60
puts "Proxy URL:       #{mask_password(proxy_url)}"
puts "Test URL:        #{test_url}"
puts "Examples:        #{to_run.size}"
puts '=' * 60
puts

passed = 0
failed = 0

to_run.each do |filename|
  name = filename.sub('.rb', '')
  print "Testing #{name}... "
  ok, err = run_example(filename)
  if ok
    puts 'OK'
    passed += 1
  else
    puts 'FAILED'
    warn "  Error: #{err}" if err
    failed += 1
  end
end

puts
puts '=' * 60
puts "Results: #{passed} passed, #{failed} failed"
puts '=' * 60

exit(failed.positive? ? 1 : 0)

File: php/run_tests.php
#!/usr/bin/env php
<?php
/**
 * Run all PHP proxy examples as tests.
 *
 * Configuration via environment variables:
 *   PROXY_URL       - Proxy URL (required), e.g., http://user:pass@proxy:8080
 *   TEST_URL        - URL to request (default: https://api.ipify.org?format=json)
 *   PROXY_HEADER    - Header name to check in response (optional)
 *   PROXY_VALUE     - Header value to send to proxy (optional)
 *
 * Usage:
 *   php run_tests.php              # Run all examples
 *   php run_tests.php guzzle       # Run specific example
 *   php run_tests.php -l           # List available examples
 *   php run_tests.php -v           # Verbose output
 */

define('EXAMPLES', [
    'curl_proxy.php' => 'cURL',
    'guzzle_proxy.php' => 'Guzzle',
    'symfony_http_client_proxy.php' => 'Symfony HttpClient',
    'buzz_proxy.php' => 'Buzz',
    'streams_proxy.php' => 'PHP Streams',
    'amphp_proxy.php' => 'Amp HTTP',
]);

function parseArgs(array $argv): array {
    $options = [
        'verbose' => false,
        'list' => false,
        'help' => false,
        'examples' => [],
    ];

    foreach (array_slice($argv, 1) as $arg) {
        switch ($arg) {
            case '-v':
            case '--verbose':
                $options['verbose'] = true;
                break;
            case '-l':
            case '--list':
                $options['list'] = true;
                break;
            case '-h':
            case '--help':
                $options['help'] = true;
                break;
            default:
                if (!str_starts_with($arg, '-')) {
                    $options['examples'][] = $arg;
                }
        }
    }

    return $options;
}

function showHelp(): void {
    echo <<<HELP
Run all PHP proxy examples as tests.

Usage:
  php run_tests.php [options] [example1] [example2] ...

Options:
  -v, --verbose    Show full output from each example
  -l, --list       List available examples
  -h, --help       Show this help message

Environment Variables:
  PROXY_URL        Proxy URL (required), e.g., http://user:pass@proxy:8080
  TEST_URL         URL to request (default: https://api.ipify.org?format=json)
  PROXY_HEADER     Header name to check in response (optional)
  PROXY_VALUE      Header value to send to proxy (optional)

Examples:
  PROXY_URL='http://proxy:8080' php run_tests.php
  php run_tests.php guzzle symfony

HELP;
}

function listExamples(): void {
    echo "Available examples:\n";
    foreach (EXAMPLES as $file => $name) {
        $key = str_replace('_proxy.php', '', $file);
        printf("  %-25s %s\n", $key, $name);
    }
}

function maskPassword(string $url): string {
    return preg_replace('/:[^:@]+@/', ':****@', $url);
}

function runExample(string $file, bool $verbose): array {
    $scriptDir = __DIR__;
    $scriptPath = "{$scriptDir}/{$file}";

    if (!file_exists($scriptPath)) {
        return ['success' => false, 'error' => "Script not found: {$scriptPath}"];
    }

    $output = [];
    $returnCode = 0;
    exec("php {$scriptPath} 2>&1", $output, $returnCode);
    $outputStr = implode("\n", $output);

    if ($verbose) {
        echo $outputStr . "\n";
    }

    return [
        'success' => $returnCode === 0,
        'output' => $outputStr,
    ];
}

function main(array $argv): int {
    $options = parseArgs($argv);

    if ($options['help']) {
        showHelp();
        return 0;
    }

    if ($options['list']) {
        listExamples();
        return 0;
    }

    $proxyUrl = getenv('PROXY_URL') ?: getenv('HTTPS_PROXY');
    if (!$proxyUrl) {
        fwrite(STDERR, "Error: Set PROXY_URL environment variable\n");
        fwrite(STDERR, "\nExample:\n");
        fwrite(STDERR, "  export PROXY_URL='http://user:pass@proxy:8080'\n");
        return 1;
    }

    $examplesToRun = EXAMPLES;
    if (!empty($options['examples'])) {
        $examplesToRun = array_filter(EXAMPLES, function($file) use ($options) {
            foreach ($options['examples'] as $arg) {
                if (str_contains($file, $arg)) {
                    return true;
                }
            }
            return false;
        }, ARRAY_FILTER_USE_KEY);
    }

    $testUrl = getenv('TEST_URL') ?: 'https://api.ipify.org?format=json';

    echo str_repeat('=', 60) . "\n";
    echo "PHP Proxy Examples - Test Runner\n";
    echo str_repeat('=', 60) . "\n";
    echo "Proxy URL:       " . maskPassword($proxyUrl) . "\n";
    echo "Test URL:        {$testUrl}\n";
    echo "Examples:        " . count($examplesToRun) . "\n";
    echo str_repeat('=', 60) . "\n\n";

    $passed = 0;
    $failed = 0;

    foreach ($examplesToRun as $file => $name) {
        echo "Testing {$name}... ";

        $result = runExample($file, $options['verbose']);

        if ($result['success']) {
            echo "OK\n";
            $passed++;
        } else {
            echo "FAILED\n";
            if (!$options['verbose']) {
                $errorLine = explode("\n", $result['output'])[0] ?? '';
                if ($errorLine) {
                    echo "  Error: {$errorLine}\n";
                }
            }
            $failed++;
        }
    }

    echo "\n" . str_repeat('=', 60) . "\n";
    echo "Results: {$passed} passed, {$failed} failed\n";
    echo str_repeat('=', 60) . "\n";

    return $failed > 0 ? 1 : 0;
}

exit(main($argv));
